crates/server/src/server/server_future.rs

```diff
@@ -642,71 +673,146 @@ impl<T: RequestHandler> ServerFuture<T> {
-
-                    // verify that the src address is safe for responses
-                    // TODO: we're relying the quinn library to actually validate responses before we get here, but this check is still worth doing
-                    if let Err(e) = sanitize_src_address(src_addr) {
-                        warn!(
-                            "address can not be responded to {src_addr}: {e}",
-                            src_addr = src_addr,
-                            e = e
-                        );
-                        continue;
-                    }
+                    },
+                    _ = drain.clone() => {
+                        // A graceful shutdown was initiated. Break out of the loop.
+                        break;
+                    },
+                };
+
+                // verify that the src address is safe for responses
+                // TODO: we're relying the quinn library to actually validate responses before we get here, but this check is still worth doing
+                if let Err(e) = sanitize_src_address(src_addr) {
+                    warn!(
+                        "address can not be responded to {src_addr}: {e}",
+                        src_addr = src_addr,
+                        e = e
+                    );
+                    continue;
+                }
 
-                    let handler = handler.clone();
-                    let dns_hostname = dns_hostname.clone();
+                let handler = handler.clone();
+                let dns_hostname = dns_hostname.clone();
 
-                    inner_join_set.spawn(async move {
-                        debug!("starting quic stream request from: {src_addr}");
+                inner_join_set.spawn(async move {
+                    debug!("starting quic stream request from: {src_addr}");
 
-                        // TODO: need to consider timeout of total connect...
-                        let result = quic_handler(handler, streams, src_addr, dns_hostname).await;
+                    // TODO: need to consider timeout of total connect...
+                    let result =
+                        quic_handler(handler, streams, src_addr, dns_hostname, drain.clone()).await;
 
-                        if let Err(e) = result {
-                            warn!("quic stream processing failed from {src_addr}: {e}")
-                        }
-                    });
-
-                    reap_tasks(&mut inner_join_set);
-                }
+                    if let Err(e) = result {
+                        warn!("quic stream processing failed from {src_addr}: {e}")
+                    }
+                });
             }
+
+            reap_tasks(&mut inner_join_set);
+            Ok(())
         });
 
         Ok(())
     }
 
-    /// This will run until a background task of the trust_dns_server ends.
+    /// Returns the drain handle used for initiating the graceful
+    /// shutdown.
+    pub fn get_drain(&self) -> Drain {
```

Comment:
Why is this needed?

Reply:
This allows the caller to use the API in a couple of different ways. They can either call `shutdown_gracefully`, which starts the shutdown and returns the shutdown future all in one shot. Alternatlively they can call `get_drain` and `block_until_done` separately, which allows the application to handle the two separately.

Reply:
I suppose my question is, why would someone want to handle the two separately?

Reply:
It's pretty common, actually. In fact, the Istio code does this for all of its server threads. If the application has its own shutdown hook (e.g. user clicks "x" on the application window, or some event is received on some channel), the shutdown hook can then trigger a drain of all the server threads. The code that triggers the shutdown would, in this case, be separate from the code that's waiting for the server to shutdown.